### PR TITLE
[MWPW-123840] Placeholders update to handle multiple locales/sheets

### PR DESCRIPTION
**Pull request by overmyheadandbody**
This updates the placeholders logic so that missing values in locales other than en-US are replaced by English labels, if they are authored. Say that we're on a ch-DE page and need a label for "_No results found_". However, the `no-results` key from the placeholders Excel document does not have a value for German, but does have one for English. So instead of returning the key name without hyphens for missing labels, such as "_no results_", we get the English value and use it, as in "_No results found_". If the English value is also missing, then the key name replacement is used. In this scenario, it might be worth logging some messages to LANA to catch such issues.

What was done:
* the `fetchPlaceholders` method now caches data for any locales that have been fetched so far, to avoid unnecessary network requests for future placeholder replacement calls;
* the `fetchPlaceholders` method can now fetch individual sheets of an Excel document. We wanted to separate concerns for authors interested in different areas of the site (general pages vs Gnav for example) and also avoid loading large sheets that might not be relevant in the current context;
* a new `replaceKeyArray` method has been exported. It can be used to replace multiple placeholders at once by passing an array of keys to the method.

Is required for: [MWPW-123840](https://jira.corp.adobe.com/browse/MWPW-123840)

**Test URLs:**
- Before:https://main--milo--adobecom.hlx.page/drafts/ramuntea/hello-milo?martech=off
- After: https://main--milo--overmyheadandbody.hlx.page/drafts/ramuntea/hello-milo?martech=off
- After for ch-DE: https://main--milo--overmyheadandbody.hlx.page/ch_de/drafts/ramuntea/hello-milo?martech=off

### CODE REVIEW

**Comment by overmyheadandbody** [comment]
Added the 'Do Not Merge' label until I fix unit tests.

**Review by mokimo — APPROVED** [review]
LGTM! Adding another unit test for this would make sense

**Review by narcis-radu — APPROVED** [review]

**Review by chrischrischris — APPROVED** [review]

**Comment by overmyheadandbody** [comment]
Hi @chrischrischris & @auniverseaway! In trying to follow the guidelines suggested in https://github.com/orgs/adobecom/teams/community/discussions/4, I was hoping you could also add the _verified_ label to the PR so I could merge it and unblock a PR from the _feds-gnav_ branch. Thanks!

**Comment by auniverseaway** [comment]
@overmyheadandbody 

Thanks for the PR. I'm confused by it, though. I'm wondering if it's that because there's not anything accompanying the PR that is using the new feature. I see a function changed, it's now exported, but it's opaque who is actually using this change, and how.

I also don't understand the caching mechanism you're referring to, either. There's one fetch request to the locale, but I don't see anything that would suggest `en-US` is cached somewhere for fallback purposes. I'm probably missing something here.

Do you think you could add some locale-based samples that show the use cases you're referring to? 

I don't know that I'm excited about offering the raw `fetchPlacehoders` function out to consumers unless there is a real need. I much prefer to keep those details internal to the module. How are `replaceKey` and `replaceText` features not working for you that would suggest you need the fetchPlaceholder function be available?

**Review by auniverseaway — CHANGES_REQUESTED** [review]
See comments above.

**Comment by overmyheadandbody** [comment]
Hey @auniverseaway! Thanks for the feedback!

Indeed, it looks confusing without actual usage. It will be used as part of the _feds-gnav_ branch, but since it touches on core Milo functionality, I thought I'd bring it up now. I'll try to summarise:

## Sheet parameter
FEDS & Gnav have a different web production team, given the large amount of content spread across locales that needs to be maintained. Having a separate FEDS-specific sheet (already implemented for [us-EN](https://adobe.sharepoint.com/:x:/r/sites/adobecom/_layouts/15/Doc.aspx?sourcedoc=%7Bda7c9a64-968c-4c93-90fe-24bcb38cc4f5%7D&action=edit&activeCell=%27Sheet1%27!B11&wdinitialsession=5d395c93-dc32-4e60-8133-03311f897ee8&wdrldsc=2&wdrldc=1&wdrldr=AccessTokenExpiredWarning%2CRefreshingExpiredAccessT&cid=2cff8858-5e70-4b2e-af99-df2a2da19052) and [ch-DE](https://adobe.sharepoint.com/:x:/r/sites/adobecom/_layouts/15/doc2.aspx?sourcedoc=%7BE50EB601-7DD1-4E40-810E-2C67F21CEFEF%7D&file=placeholders.xlsx&action=default&mobileredirect=true&cid=00787479-df1e-466a-b47f-050cc4a070cc)) allows for more focus on relevant areas for authors, should avoid unintended cross-platform edits of similar labels and should be good for overall maintainability of sheets; from a performance standpoint, this should also be beneficial, since calls are more focused on useful data for the context.

## Locales and fallback mechanism
The current approach in Milo is that if a label for a placeholder doesn't exist, then it's replaced with the key name. In reality, we usually use short key names for large content, so this approach is not always ideal. For example the value for `user-help` is _Please try again selecting fewer filters or clearing all selections._. So, currently, if the _ch-DE_ placeholders sheet doesn't have a label for `user-help`, then it will be replaced with _user help_.

English is the first locale populated by authors, so we could assume most placeholders will have an English label available. That's why, in the Gnav on Milo refactor, we're implementing a slightly different placeholders approach: if the _ch-DE_ placeholders sheet doesn't have a label for `user-help`, fetch the sheet for _us-EN_ and get the content from there.

Since there might be multiple such placeholders, we of course don't want to perform any additional network requests, so all of the already fetched locales and sheets are stored in the `placeholders` object.

## Quick example
_us-EN_ sheet has these placeholders: search|Search, sign-in|Sign In;
_ch-DE_ sheet has these placeholder: search|Suchen;

A new method is defined in Gnav utilities, `getPlaceholders`, which would be called with an array of keys to be fetched.

```
// current locale is us-EN;
// both labels are available,
// Milo fetchPlaceholders method is called,
// caches labels under the 'us-EN' property
getPlaceholders(['search', 'sign-in']); // result is ['Search', 'Sign In']

// current locale is ch-DE;
// just one label is available,
// Milo fetchPlaceholders method is called to get the German labels,
// caches them under the 'ch-DE' property;
// Milo fetchPlaceholders method is called again to get the English labels,
// caches them under the 'us-EN' property
getPlaceholders(['search', 'sign-in']); // result is ['Suchen', 'Sign In']
```

## Why `replaceText` is not enough
The method assumes all placeholders have a label available in a single language. As stated previously, different placeholders may have either locale-specific labels or a default English one.

## Alternative approach
Since FEDS & Gnav will use a separate sheet in the placeholders document anyway, thus making a different request, we could have our own method to fetch placeholders. However, I was trying to avoid code duplication and thought that other teams might want to use other sheets and have multiple locale translations available if needed. If that's not the case, we can definitely contain this approach just to FEDS. If this **is** the case, I could potentially update the `replaceText` and `replaceKey` methods in Milo to reflect our _us-EN_ locale fallback approach, but I fear this might lead to a bit of friction 🙂

File: libs/features/placeholders.js
Comment (overmyheadandbody) [inline libs/features/placeholders.js:36]: I'm not really sure this is the correct approach, I'm not aware of the various use-cases for `contentRoot`.

Comment (mokimo) [reply libs/features/placeholders.js:36]: Hmm, this is a tricky one.

One scenario where this would fail is if we used placeholders for the Profile which can be used standalone, I would guess this wouldn't work anymore then.

I'd say it's fine to solve that once there's a use-case though

**Review by mokimo — APPROVED** [review]
We should also write some unit tests for the new functionality. The placeholders are pretty core - so we should be solid on the tests there to avoid future regressions on changes

**Review by narcis-radu — APPROVED** [review]

**Comment by overmyheadandbody** [comment]
Hi @auniverseaway! Could we get any additional feedback on this, please? Or, if it's good to go, could you add the appropriate labels and merge it? Thanks!

**Comment by overmyheadandbody** [comment]
Hi @auniverseaway! Could we get any additional feedback on this, please? Or, if it's good to go, could you add the appropriate labels and merge it? Thanks!

File: libs/features/placeholders.js
Comment (vhargrave) [reply libs/features/placeholders.js:36]: @overmyheadandbody you're spot on that the origin can be different from the contentRoot. In [stock](https://github.com/adobecom/stock/blob/3c631b44403019087606321c4cd4a28e428e9834/pages/scripts/scripts.js#L28) for example, the contentRoot is under pages

File: libs/features/placeholders.js
Comment (vhargrave) [inline libs/features/placeholders.js:30]: I'd rename to getDefaultPlaceholder? I was kind of expecting it to return an array of placeholders otherwise.

**Review by vhargrave — CHANGES_REQUESTED** [review]
Works really well. The only big thing I noticed was something you'd already mentioned, and that's if you should use the contentRoot instead of the origin. Yes, you should :) 
Found two other super small things, and then it's looking pretty good imo. 
Since placeholders is a bigger deal I would feel safer if you had an example test of it running on bacom and stock. Stock is where you would have noticed the contentRoot issue btw.

File: libs/features/placeholders.js
Comment (mokimo) [reply libs/features/placeholders.js:40]: In theory I'd agree with that, but historical evidence is that things are just forgotten often for non-English locales which is something we encountered often enough in production in the dexter/aem world.

It's a large number of pages and content so IMO it's better if the user sees something understandable. (Assuming english is more understandable )

Comment (overmyheadandbody) [reply libs/features/placeholders.js:40]: If we assume a content producer is able to spot an English key name on a German locale, as is the current case with missing placeholders, I think it's safe to assume they would also be able to spot an English word/sentence on a German page. So if content producers pay attention to these details, we have nothing to worry about. However, if this detail is missed, then at least the end user gets some relevant information, albeit in another (international) language. In the future, we were thinking to include some LANA logging when placeholders use fallbacks so we see where translations are missing.

Comment (vhargrave) [reply libs/features/placeholders.js:40]: I'm not sure I agree that users have the same chance of spotting keys vs translated text: 
Here's are two example pages I made. Open the first link and try to see the English text. Then open the second link: 
https://main--milo--adobecom.hlx.page/drafts/vhargrave/normaltext
https://main--milo--adobecom.hlx.page/drafts/vhargrave/keytext

Maybe we can extend this in a later task to still show the keys on stage, but show the backup text on prod? That way we get the best of both worlds. What do you guys think?

Comment (narcis-radu) [reply libs/features/placeholders.js:40]: Instead of showing different content on STAGE vs PROD, why don't we just highlight backup text when a given query param or global object is set? We would be able to use it in our automated tests and output all issues. @overmyheadandbody's suggestion with LANA logging is another great way of capturing issues and fixing them asap.

Comment (vhargrave) [reply libs/features/placeholders.js:40]: That's a great idea .
And yeah, I forgot to write it, but the Lana logging would also be a super addition.

File: libs/features/placeholders.js
Comment (vhargrave) [reply libs/features/placeholders.js:32]: @narcis-radu 
aha sorry, just use the contentRoot from the config object. 
You're already passing in the config object into the method, so when creating your default config for example, you could do this : 
```
const origin = config.origin || window.location.origin;
const defaultContentRoot = `${origin}${config.contentRoot ?? ''}`;
const defaultConfig = {
  locale: {
    ietf: defaultLocale,
    contentRoot: defaultContentRoot,
  },
};
```
This is probably the simplest approach. I'll ping if I have a different idea.

Comment (vhargrave) [reply libs/features/placeholders.js:32]: If you want to see an example of it being used in a project btw, there's one in [stock](https://github.com/adobecom/stock/blob/3c631b44403019087606321c4cd4a28e428e9834/pages/scripts/scripts.js#L28)

File: libs/features/placeholders.js
Comment (overmyheadandbody) [reply libs/features/placeholders.js:36]: Nice catch, the method was returning either a set of placeholders or just one, which was misleading. The purpose of this method is to return a whole set, the name has been kept. However, other pieces of logic have been updated accordingly.

File: libs/features/placeholders.js
Comment (overmyheadandbody) [reply libs/features/placeholders.js:44]: Although that particular example from Victor's test tries to prove a point, I'd argue that the effect is due to the brackets and dashes. This is not how things behave today in Milo.

Today, if a label is not found for the `you-saw-this-quicker-right` label, it gets replaced with `you saw this quicker right`, not with `[[you-saw-this-quicker-right]]` as that page might suggest. So it's exactly the same when trying to spot a missed translation.

Comment (vhargrave) [reply libs/features/placeholders.js:44]: Yes, that's true. My dream scenario would also include not removing the brackets from the placeholder, so they stick out. 
Technically fine for me regarding your proposal, but one of the admins (chris m., ryan, chris p.) has to merge, since they're admins.

**Comment by overmyheadandbody** [comment]
Had a chat with @vhargrave who offered some details on different use-cases for the content root. Made the appropriate changes, things should be working now.

**Review by vhargrave — APPROVED** [review]
Looks good - nice job!

**Review by auniverseaway — APPROVED** [review]

**Comment by auniverseaway** [comment]
@narcis-radu what do you want to do about this with your WIP work? I would prefer we keep this and all placeholder work stays out of gnav.

**Comment by overmyheadandbody** [comment]
Hi, @auniverseaway! This PR is not WIP, you can go ahead and merge it, it's been laying around for quite some time. The placeholders work that has been done in the gnav feature branch has two layers:
1. An exact copy-paste of the code from this PR so we can move forward with our development
2. Further exploration of some ideas to fetch additional data from placeholder sheets. We likely don't have a good use-case for that extra data for now, so the code will be reverted to how it is described in 1.

Thus, the code from this PR will be the single source of placeholder truth.

This being said, we will have to provide another config when using these methods, since the Gnav content should be centralised in one place, regardless of the consumer and their particular set of placeholders. This will be addressed in the Gnav feature branch.

**Comment by SilviuLCF** [comment]
Hi @auniverseaway,
Regression for the placeholders logic was performed and things are working as expected.
We can go ahead with merging the PR. 
I can’t add the “verified” label myself, how could I get permission to do this?

Thx

**Comment by mokimo** [comment]
I added the verified label for now until you get the correct permissions @SilviuLCF

**Pull request merged**
